File: tests/vp/test_mpp_decoder.cpp
#include "module/vi/module_fileReader.hpp"
#include "module/vp/module_mppdec.hpp"
#include "tests/media_channel_dump.hpp"

#include <atomic>
#include <chrono>
#include <unistd.h>

using namespace FFMedia;
using namespace std;

std::atomic_bool is_exit(false);

void output_callback(const string& name, int queue_size,
                     shared_ptr<MediaBuffer> buffer)
{
    static uint64_t frame_count;
    static auto start_time = std::chrono::high_resolution_clock::now();
    if (++frame_count % 100 == 0) {
        auto diff = std::chrono::duration_cast<std::chrono::milliseconds>(
                        std::chrono::high_resolution_clock::now() - start_time)
                        .count();
        ff_info("%s: queue %d, decoded %ld frames, %ld ms, %ld fps\n",
                name.c_str(), queue_size, frame_count, diff,
                frame_count * 1000 / diff);
    }
}

void status_change_callback(const string& name, MediaStatus status)
{
    ff_info("%s status changed to %d\n", name.c_str(), static_cast<int>(status));
    if (status == MediaStatus::EOS)
        is_exit = true;
}

int main(int argc, char** argv)
{
    int ret;

    if (argc < 2) {
        ff_error("\nUsage: %s file\n", argv[0]);
        return -1;
    };

    // create file reader module
    auto f_reader = make_shared<ModuleFileReader>(argv[1]);
    /* Set the callback function for the reader */
    f_reader->setMediaStatusChangeHooker(status_change_callback);
    ret = f_reader->init();
    if (ret < 0) {
        ff_error("Failed to init file reader, %d\n", ret);
        return ret;
    }

    // Print stream information through the reader output channels.
    dumpOutputMediaChannels(*f_reader);

    /* Create a mpp decoder module. */
    auto v_dec = make_shared<ModuleMppDec>();
    ret = v_dec->connectProducer(f_reader);
    if (ret < 0) {
        ff_error("Failed to connect file reader to mpp decoder, %d\n", ret);
        return ret;
    }
    v_dec->setBufferCount(20);
    v_dec->setMediaBufferProduceHooker(output_callback);
    ret = v_dec->init();
    if (ret < 0) {
        ff_error("Failed to init mpp decoder, %d\n", ret);
        return ret;
    }
    dumpOutputMediaChannels(*v_dec);

    ff_info("\n========================================================\n\n");

    // start
    f_reader->start();
    while (!is_exit) {
        sleep(1);
    }
    f_reader->stop();
    ff_info("Exit.\n");
    return 0;
}

File: tests/vp/test_mpp_decoder.py
#!/usr/bin/env python3

import sys
import time
import ff_pymedia as ff

frame_counter = 0
init_time = time.time()

def output_callback(name, queue_size, buffer):
    global frame_counter, init_time

    frame_counter += 1

    if frame_counter == 1:
        init_time = time.time()

    if frame_counter % 100 == 0:
        duration = time.time() - init_time
        fps = frame_counter / duration
        print(
            f"{name}: queue {queue_size}, decoded {frame_counter} frames, "
            f"{duration:.3f}s, {fps:.2f} fps"
        )

    # get memoryview object
    #data = buffer.getActiveData()
    #print("Data size ", data.nbytes)

def status_change_callback(name, status):
    print(f"{name} module state has changed: ", status)


def dump_output_channels(module):
    print(f"{module.getName()} output channels:")
    for channel in module.getOutputMediaChannels():
        extra_size = channel.extra_data.getActiveSize() if channel.extra_data else 0
        if channel.media_type == ff.BUFFER_TYPE_VIDEO:
            image = channel.image_para
            print(
                f"  Channel[{channel.id}] {channel.name}: video, "
                f"codec {channel.codec}, format {image.v4l2Fmt}, "
                f"size {image.width}x{image.height}, "
                f"stride {image.hstride}x{image.vstride}, extra {extra_size} bytes"
            )
        elif channel.media_type == ff.BUFFER_TYPE_AUDIO:
            sample = channel.sample_info
            print(
                f"  Channel[{channel.id}] {channel.name}: audio, "
                f"codec {channel.codec}, format {sample.fmt}, "
                f"channels {sample.channels}, sample rate {sample.sample_rate}, "
                f"samples {sample.nb_samples}, extra {extra_size} bytes"
            )
        else:
            print(
                f"  Channel[{channel.id}] {channel.name}: "
                f"type {channel.media_type}, codec {channel.codec}, "
                f"extra {extra_size} bytes"
            )

def main():
    if len(sys.argv) > 1:
        file_path = sys.argv[1]
    else:
        file_path = input("Please enter the file path or url: ")

    # create FFmpeg demux module.
    demuxer = ff.ModuleFFmpegDemux(file_path, -1)
    demuxer.setFormatOption("probesize", "200K", 0)
    # Set the callback function for the demuxer.
    demuxer.setMediaStatusChangeHooker(status_change_callback)
    ret = demuxer.init()
    if ret < 0:
        print("Failed to init demuxer")
        return 1

    # View stream information through the demuxer output channels.
    dump_output_channels(demuxer)

    # Create a mpp decoder module.
    v_decoder = ff.ModuleMppDec(demuxer.getOutputImagePara())
    ret = v_decoder.connectProducer(demuxer)
    if ret < 0:
        print("Failed to connect demuxer to mpp decoder, \t", ret)
        return ret
    # Set the callback function for the decoder.
    v_decoder.setMediaBufferProduceHooker(output_callback)
    ret = v_decoder.init()
    if ret < 0:
        print("Failed to init mpp decoder, \t", ret)
        return ret

    dump_output_channels(v_decoder)


    print("\n============================START=============================\n")
    demuxer.start()
    input("wait...")
    demuxer.stop()
    print("\n============================STOP==============================\n")

if __name__ == "__main__":
    main()
